File: Source/AlphaTab/Rendering/Glyphs/ScoreBeatGlyph.cs
/*
 * This file is part of alphaTab.
 * Copyright © 2017, Daniel Kuschny and Contributors, All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or at your option any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

using System;
using AlphaTab.Collections;
using AlphaTab.Model;
using AlphaTab.Platform;

namespace AlphaTab.Rendering.Glyphs
{
    public class ScoreBeatGlyph : BeatOnNoteGlyphBase
    {
        public ScoreNoteChordGlyph NoteHeads { get; set; }
        public ScoreRestGlyph RestGlyph { get; set; }

        public override void UpdateBeamingHelper()
        {
            if (NoteHeads != null)
            {
                NoteHeads.UpdateBeamingHelper(Container.X + X);
            }
            else if (RestGlyph != null)
            {
                RestGlyph.UpdateBeamingHelper(Container.X + X);
            }
        }

        public override void DoLayout()
        {
            // create glyphs
            var sr = (ScoreBarRenderer)Renderer;
            if (!Container.Beat.IsEmpty)
            {
                if (!Container.Beat.IsRest)
                {
                    //
                    // Note heads
                    //
                    NoteHeads = new ScoreNoteChordGlyph();
                    NoteHeads.Beat = Container.Beat;
                    NoteHeads.BeamingHelper = BeamingHelper;
                    foreach (var note in Container.Beat.Notes)
                    {
                        CreateNoteGlyph(note);
                    }
                    AddGlyph(NoteHeads);

                    //
                    // Note dots
                    //
                    if (Container.Beat.Dots > 0)
                    {
                        AddGlyph(new SpacingGlyph(0, 0, 5 * Scale));
                        for (var i = 0; i < Container.Beat.Dots; i++)
                        {
                            var group = new GlyphGroup(0, 0);
                            foreach (var note in Container.Beat.Notes)
                            {
                                CreateBeatDot(sr.GetNoteLine(note), group);
                            }
                            AddGlyph(group);
                        }
                    }
                }
                else
                {
                    var dotLine = 0;
                    var line = 0;
                    var offset = 0;

                    switch (Container.Beat.Duration)
                    {
                        case Duration.QuadrupleWhole:
                            line = 6;
                            dotLine = 5;
                            break;
                        case Duration.DoubleWhole:
                            line = 6;
                            dotLine = 5;
                            break;
                        case Duration.Whole:
                            line = 4;
                            dotLine = 5;
                            break;
                        case Duration.Half:
                            line = 6;
                            dotLine = 5;
                            break;
                        case Duration.Quarter:
                            line = 6;
                            offset = -2;
                            dotLine = 5;
                            break;
                        case Duration.Eighth:
                            line = 6;
                            dotLine = 5;
                            break;
                        case Duration.Sixteenth:
                            line = 6;
                            dotLine = 5;
                            break;
                        case Duration.ThirtySecond:
                            line = 6;
                            dotLine = 3;
                            break;
                        case Duration.SixtyFourth:
                            line = 6;
                            dotLine = 3;
                            break;
                        case Duration.OneHundredTwentyEighth:
                            line = 6;
                            dotLine = 3;
                            break;
                        case Duration.TwoHundredFiftySixth:
                            line = 6;
                            dotLine = 3;
                            break;
                    }

                    var y = sr.GetScoreY(line, offset);

                    RestGlyph = new ScoreRestGlyph(0, y, Container.Beat.Duration);
                    RestGlyph.Beat = Container.Beat;
                    RestGlyph.BeamingHelper = BeamingHelper;
                    AddGlyph(RestGlyph);

                    //
                    // Note dots
                    //
                    if (Container.Beat.Dots > 0)
                    {
                        AddGlyph(new SpacingGlyph(0, 0, 5 * Scale));
                        for (var i = 0; i < Container.Beat.Dots; i++)
                        {
                            var group = new GlyphGroup(0, 0);
                            CreateBeatDot(dotLine, group);
                            AddGlyph(group);
                        }
                    }
                }
            }

            base.DoLayout();
        }

        private void CreateBeatDot(int line, GlyphGroup group)
        {
            var sr = (ScoreBarRenderer)Renderer;
            group.AddGlyph(new CircleGlyph(0, sr.GetScoreY(line), 1.5f * Scale));
        }

        private static readonly FastDictionary<int, bool> NormalKeys;
        private static readonly FastDictionary<int, bool> XKeys;

        static ScoreBeatGlyph()
        {
            // ReSharper disable ForCanBeConvertedToForeach
            NormalKeys = new FastDictionary<int, bool>();
            var normalKeyNotes = new[] { 32, 34, 35, 36, 38, 39, 40, 41, 43, 45, 47, 48, 50, 55, 56, 58, 60, 61 };
            for (int i = 0; i < normalKeyNotes.Length; i++)
            {
                NormalKeys[normalKeyNotes[i]] = true;
            }
            XKeys = new FastDictionary<int, bool>();
            var xKeyNotes = new[] { 31, 33, 37, 42, 44, 54, 62, 63, 64, 65, 66 };
            for (int i = 0; i < xKeyNotes.Length; i++)
            {
                XKeys[xKeyNotes[i]] = true;
            }
            // ReSharper restore ForCanBeConvertedToForeach
        }

        private Glyph CreateNoteHeadGlyph(Note n)
        {
            var isGrace = Container.Beat.GraceType != GraceType.None;
            if (n.Beat.Voice.Bar.Staff.Track.IsPercussion)
            {
                var value = n.RealValue;

                if (value <= 30 || value >= 67 || NormalKeys.ContainsKey(value))
                {
                    return new NoteHeadGlyph(0, 0, Duration.Quarter, isGrace);
                }
                if (XKeys.ContainsKey(value))
                {
                    return new DrumSticksGlyph(0, 0, isGrace);
                }
                if (value == 46)
                {
                    return new HiHatGlyph(0, 0, isGrace);
                }
                if (value == 49 || value == 57)
                {
                    return new DiamondNoteHeadGlyph(0, 0, isGrace);
                }
                if (value == 52)
                {
                    return new ChineseCymbalGlyph(0, 0, isGrace);
                }
                if (value == 51 || value == 53 || value == 59)
                {
                    return new RideCymbalGlyph(0, 0, isGrace);
                }
                return new NoteHeadGlyph(0, 0, Duration.Quarter, isGrace);
            }
            if (n.IsDead)
            {
                return new DeadNoteHeadGlyph(0, 0, isGrace);
            }
            if (n.HarmonicType == HarmonicType.None)
            {
                return new NoteHeadGlyph(0, 0, n.Beat.Duration, isGrace);
            }
            return new DiamondNoteHeadGlyph(0, 0, isGrace);
        }

        private void CreateNoteGlyph(Note n)
        {
            var sr = (ScoreBarRenderer)Renderer;
            var noteHeadGlyph = CreateNoteHeadGlyph(n);

            // calculate y position
            var line = sr.GetNoteLine(n);

            noteHeadGlyph.Y = sr.GetScoreY(line);
            NoteHeads.AddNoteGlyph(noteHeadGlyph, n, line);

            if (n.IsStaccato && !NoteHeads.BeatEffects.ContainsKey("Staccato"))
            {
                NoteHeads.BeatEffects["Staccato"] = new CircleGlyph(0, 0, 1.5f);
            }

            if (n.Accentuated == AccentuationType.Normal && !NoteHeads.BeatEffects.ContainsKey("Accent"))
            {
                NoteHeads.BeatEffects["Accent"] = new AccentuationGlyph(0, 0, AccentuationType.Normal);
            }
            if (n.Accentuated == AccentuationType.Heavy && !NoteHeads.BeatEffects.ContainsKey("HAccent"))
            {
                NoteHeads.BeatEffects["HAccent"] = new AccentuationGlyph(0, 0, AccentuationType.Heavy);
            }
        }
    }
}

File: src/rendering/glyphs/ScoreBeatGlyph.ts
import { AccentuationType } from '@src/model/AccentuationType';
import { Duration } from '@src/model/Duration';
import { GraceType } from '@src/model/GraceType';
import { HarmonicType } from '@src/model/HarmonicType';
import { type Note, NoteSubElement } from '@src/model/Note';
import { AccentuationGlyph } from '@src/rendering/glyphs/AccentuationGlyph';
import { BeatOnNoteGlyphBase } from '@src/rendering/glyphs/BeatOnNoteGlyphBase';
import { CircleGlyph } from '@src/rendering/glyphs/CircleGlyph';
import { DeadNoteHeadGlyph } from '@src/rendering/glyphs/DeadNoteHeadGlyph';
import { DiamondNoteHeadGlyph } from '@src/rendering/glyphs/DiamondNoteHeadGlyph';
import { GhostNoteContainerGlyph } from '@src/rendering/glyphs/GhostNoteContainerGlyph';
import { GlyphGroup } from '@src/rendering/glyphs/GlyphGroup';
import { NoteHeadGlyph } from '@src/rendering/glyphs/NoteHeadGlyph';
import { ScoreNoteChordGlyph } from '@src/rendering/glyphs/ScoreNoteChordGlyph';
import { ScoreRestGlyph } from '@src/rendering/glyphs/ScoreRestGlyph';
import { ScoreWhammyBarGlyph } from '@src/rendering/glyphs/ScoreWhammyBarGlyph';
import { SpacingGlyph } from '@src/rendering/glyphs/SpacingGlyph';
import type { ScoreBarRenderer } from '@src/rendering/ScoreBarRenderer';
import type { NoteXPosition, NoteYPosition } from '@src/rendering/BarRendererBase';
import type { BeatBounds } from '@src/rendering/utils/BeatBounds';
import { PercussionMapper } from '@src/model/PercussionMapper';
import { PercussionNoteHeadGlyph } from '@src/rendering/glyphs/PercussionNoteHeadGlyph';
import { Logger } from '@src/Logger';
import { ArticStaccatoAboveGlyph } from '@src/rendering/glyphs/ArticStaccatoAboveGlyph';
import { MusicFontSymbol } from '@src/model/MusicFontSymbol';
import { type ICanvas, TextBaseline } from '@src/platform/ICanvas';
import { PictEdgeOfCymbalGlyph } from '@src/rendering/glyphs/PictEdgeOfCymbalGlyph';
import { PickStrokeGlyph } from '@src/rendering/glyphs/PickStrokeGlyph';
import { PickStroke } from '@src/model/PickStroke';
import { GuitarGolpeGlyph } from '@src/rendering/glyphs/GuitarGolpeGlyph';
import { BeamingHelper } from '@src/rendering/utils/BeamingHelper';
import { StringNumberContainerGlyph } from '@src/rendering/glyphs/StringNumberContainerGlyph';
import { SlashNoteHeadGlyph } from '@src/rendering/glyphs/SlashNoteHeadGlyph';
import { BeatSubElement } from '@src/model/Beat';
import { ElementStyleHelper } from '@src/rendering/utils/ElementStyleHelper';
import type { MusicFontGlyph } from '@src/rendering/glyphs/MusicFontGlyph';

export class ScoreBeatGlyph extends BeatOnNoteGlyphBase {
    private _collisionOffset: number = -1000;
    private _skipPaint: boolean = false;

    public slash: SlashNoteHeadGlyph | null = null;
    public noteHeads: ScoreNoteChordGlyph | null = null;
    public restGlyph: ScoreRestGlyph | null = null;

    protected override get effectElement() {
        return BeatSubElement.StandardNotationEffects;
    }

    public override getNoteX(note: Note, requestedPosition: NoteXPosition): number {
        return this.noteHeads ? this.noteHeads.getNoteX(note, requestedPosition) : 0;
    }

    public override buildBoundingsLookup(beatBounds: BeatBounds, cx: number, cy: number) {
        if (this.noteHeads) {
            this.noteHeads.buildBoundingsLookup(beatBounds, cx + this.x, cy + this.y);
        }
    }

    public override getNoteY(note: Note, requestedPosition: NoteYPosition): number {
        return this.noteHeads ? this.noteHeads.getNoteY(note, requestedPosition) : 0;
    }

    public override updateBeamingHelper(): void {
        if (this.noteHeads) {
            this.noteHeads.updateBeamingHelper(this.container.x + this.x);
        } else if (this.restGlyph) {
            this.restGlyph.updateBeamingHelper(this.container.x + this.x);
            if (this.renderer.bar.isMultiVoice && this._collisionOffset === -1000) {
                this._collisionOffset = this.renderer.helpers.collisionHelper.applyRestCollisionOffset(
                    this.container.beat,
                    this.restGlyph.y,
                    (this.renderer as ScoreBarRenderer).getScoreHeight(1)
                );
                this.y += this._collisionOffset;
                const existingRests = this.renderer.helpers.collisionHelper.restDurationsByDisplayTime;
                if (
                    existingRests.has(this.container.beat.playbackStart) &&
                    existingRests.get(this.container.beat.playbackStart)!.has(this.container.beat.playbackDuration) &&
                    existingRests.get(this.container.beat.playbackStart)!.get(this.container.beat.playbackDuration) !==
                        this.container.beat.id
                ) {
                    this._skipPaint = true;
                }
            }
        } else if (this.slash) {
            this.slash.updateBeamingHelper(this.container.x + this.x);
        }
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        if (!this._skipPaint) {
            super.paint(cx, cy, canvas);
        }
    }

    public override doLayout(): void {
        // create glyphs
        const sr: ScoreBarRenderer = this.renderer as ScoreBarRenderer;
        if (!this.container.beat.isEmpty) {
            if (!this.container.beat.isRest) {
                if (this.container.beat.slashed) {
                    const isGrace = this.container.beat.graceType !== GraceType.None;
                    const line = (sr.heightLineCount - 1) / 2;
                    const slashY = sr.getLineY(line);
                    const slashNoteHead = new SlashNoteHeadGlyph(
                        0,
                        slashY,
                        this.container.beat.duration,
                        isGrace,
                        this.container.beat
                    );
                    slashNoteHead.noteHeadElement = NoteSubElement.StandardNotationNoteHead;
                    slashNoteHead.effectElement = BeatSubElement.StandardNotationEffects;
                    this.slash = slashNoteHead;
                    slashNoteHead.beat = this.container.beat;
                    slashNoteHead.beamingHelper = this.beamingHelper;
                    this.addNormal(slashNoteHead);
                } else {
                    //
                    // Note heads
                    //
                    const noteHeads = new ScoreNoteChordGlyph();
                    this.noteHeads = noteHeads;
                    noteHeads.beat = this.container.beat;
                    noteHeads.beamingHelper = this.beamingHelper;
                    const ghost: GhostNoteContainerGlyph = new GhostNoteContainerGlyph(false);
                    ghost.renderer = this.renderer;
                    for (const note of this.container.beat.notes) {
                        if (note.isVisible) {
                            this.createNoteGlyph(note);
                            ghost.addParenthesis(note);
                        }
                    }
                    this.addNormal(noteHeads);
                    if (!ghost.isEmpty) {
                        this.addNormal(
                            new SpacingGlyph(
                                0,
                                0,
                                4 * (this.container.beat.graceType !== GraceType.None ? NoteHeadGlyph.GraceScale : 1)
                            )
                        );
                        this.addEffect(ghost);
                    }
                }

                //
                // Whammy Bar
                if (this.container.beat.hasWhammyBar) {
                    const whammy: ScoreWhammyBarGlyph = new ScoreWhammyBarGlyph(this.container.beat);
                    whammy.renderer = this.renderer;
                    whammy.doLayout();
                    this.container.ties.push(whammy);
                }
                //
                // Note dots
                //
                if (this.container.beat.dots > 0) {
                    this.addNormal(new SpacingGlyph(0, 0, 5));
                    for (let i: number = 0; i < this.container.beat.dots; i++) {
                        const group: GlyphGroup = new GlyphGroup(0, 0);
                        group.renderer = this.renderer;
                        for (const note of this.container.beat.notes) {
                            const g = this.createBeatDot(sr.getNoteLine(note), group);
                            g.colorOverride = ElementStyleHelper.noteColor(
                                sr.resources,
                                NoteSubElement.StandardNotationEffects,
                                note
                            );
                        }
                        this.addEffect(group);
                    }
                }
            } else {
                let line = Math.ceil((this.renderer.bar.staff.standardNotationLineCount - 1) / 2) * 2;

                // this positioning is quite strange, for most staff line counts
                // the whole/rest are aligned as half below the whole rest.
                // but for staff line count 1 and 3 they are aligned centered on the same line.
                if (
                    this.container.beat.duration === Duration.Whole &&
                    this.renderer.bar.staff.standardNotationLineCount !== 1 &&
                    this.renderer.bar.staff.standardNotationLineCount !== 3
                ) {
                    line -= 2;
                }

                const restGlyph = new ScoreRestGlyph(0, sr.getScoreY(line), this.container.beat.duration);
                this.restGlyph = restGlyph;
                restGlyph.beat = this.container.beat;
                restGlyph.beamingHelper = this.beamingHelper;
                this.addNormal(restGlyph);

                if (this.renderer.bar.isMultiVoice) {
                    if (this.container.beat.voice.index === 0) {
                        const restSizes = BeamingHelper.computeLineHeightsForRest(this.container.beat.duration);
                        const restTop = restGlyph.y - sr.getScoreHeight(restSizes[0]);
                        const restBottom = restGlyph.y + sr.getScoreHeight(restSizes[1]);
                        this.renderer.helpers.collisionHelper.reserveBeatSlot(this.container.beat, restTop, restBottom);
                    } else {
                        this.renderer.helpers.collisionHelper.registerRest(this.container.beat);
                    }
                }

                if (this.beamingHelper) {
                    this.beamingHelper.applyRest(this.container.beat, line);
                }

                //
                // Note dots
                //
                if (this.container.beat.dots > 0) {
                    this.addNormal(new SpacingGlyph(0, 0, 5));
                    for (let i: number = 0; i < this.container.beat.dots; i++) {
                        const group: GlyphGroup = new GlyphGroup(0, 0);
                        group.renderer = this.renderer;
                        this.createBeatDot(line, group);
                        this.addEffect(group);
                    }
                }
            }
        }
        super.doLayout();
        if (this.container.beat.isEmpty) {
            this.centerX = this.width / 2;
        } else if (this.restGlyph) {
            this.centerX = this.restGlyph!.x + this.restGlyph!.width / 2;
        } else if (this.noteHeads) {
            this.centerX = this.noteHeads!.x + this.noteHeads!.width / 2;
        } else if (this.slash) {
            this.centerX = this.slash!.x + this.slash!.width / 2;
        }
    }

    private createBeatDot(line: number, group: GlyphGroup) {
        const sr: ScoreBarRenderer = this.renderer as ScoreBarRenderer;
        const g = new CircleGlyph(0, sr.getScoreY(line), 1.5);
        group.addGlyph(g);
        return g;
    }

    private createNoteHeadGlyph(n: Note): MusicFontGlyph {
        const isGrace: boolean = this.container.beat.graceType !== GraceType.None;

        const style = n.style;
        if (style?.noteHead !== undefined) {
            const noteHead = new NoteHeadGlyph(0, 0, n.beat.duration, isGrace);
            // NOTE: sizes are not yet perfect
            // will be done in https://github.com/CoderLine/alphaTab/issues/1949
            noteHead.symbol = style!.noteHead!;
            if (style.noteHeadCenterOnStem) {
                noteHead.centerOnStem = true;
            }
            return noteHead;
        }

        // TODO: here we should unify it to one common glyph which knows all sizes.

        if (n.beat.voice.bar.staff.isPercussion) {
            const articulation = PercussionMapper.getArticulation(n);
            if (articulation) {
                return new PercussionNoteHeadGlyph(0, 0, articulation, n.beat.duration, isGrace);
            }
            Logger.warning('Rendering', `No articulation found for percussion instrument ${n.percussionArticulation}`);
        }
        if (n.isDead) {
            return new DeadNoteHeadGlyph(0, 0, isGrace);
        }
        if (n.beat.graceType === GraceType.BendGrace) {
            return new NoteHeadGlyph(0, 0, Duration.Quarter, true);
        }
        if (n.harmonicType === HarmonicType.Natural) {
            return new DiamondNoteHeadGlyph(0, 0, n.beat.duration, isGrace);
        }

        return new NoteHeadGlyph(0, 0, n.beat.duration, isGrace);
    }

    private createNoteGlyph(n: Note): void {
        if (n.beat.graceType === GraceType.BendGrace && !n.hasBend) {
            return;
        }
        const sr: ScoreBarRenderer = this.renderer as ScoreBarRenderer;
        const noteHeadGlyph = this.createNoteHeadGlyph(n);
        noteHeadGlyph.colorOverride = ElementStyleHelper.noteColor(
            sr.resources,
            NoteSubElement.StandardNotationNoteHead,
            n
        );

        // calculate y position
        let line: number = sr.getNoteLine(n);
        noteHeadGlyph.y = sr.getScoreY(line);
        this.noteHeads!.addNoteGlyph(noteHeadGlyph, n, line);
        if (n.harmonicType !== HarmonicType.None && n.harmonicType !== HarmonicType.Natural) {
            // create harmonic note head.
            const harmonicFret: number = n.displayValue + n.harmonicPitch;
            const harmonicsGlyph = new DiamondNoteHeadGlyph(
                0,
                0,
                n.beat.duration,
                this.container.beat.graceType !== GraceType.None
            );
            harmonicsGlyph.colorOverride = noteHeadGlyph.colorOverride;
            line = sr.accidentalHelper.getNoteLineForValue(harmonicFret, false);
            harmonicsGlyph.y = sr.getScoreY(line);
            this.noteHeads!.addNoteGlyph(harmonicsGlyph, n, line);
        }

        const belowBeatEffects = this.noteHeads!.belowBeatEffects;
        const aboveBeatEffects = this.noteHeads!.aboveBeatEffects;

        if (n.isStaccato && !belowBeatEffects.has('Staccato')) {
            belowBeatEffects.set('Staccato', new ArticStaccatoAboveGlyph(0, 0));
        }
        if (n.accentuated === AccentuationType.Normal && !belowBeatEffects.has('Accent')) {
            belowBeatEffects.set('Accent', new AccentuationGlyph(0, 0, n));
        }
        if (n.accentuated === AccentuationType.Heavy && !belowBeatEffects.has('HAccent')) {
            belowBeatEffects.set('HAccent', new AccentuationGlyph(0, 0, n));
        }
        if (n.accentuated === AccentuationType.Tenuto && !belowBeatEffects.has('Tenuto')) {
            belowBeatEffects.set('Tenuto', new AccentuationGlyph(0, 0, n));
        }
        if (n.showStringNumber && n.isStringed) {
            let container: StringNumberContainerGlyph;
            if (!aboveBeatEffects.has('StringNumber')) {
                container = new StringNumberContainerGlyph(0, 0);
                aboveBeatEffects.set('StringNumber', container);
            } else {
                container = aboveBeatEffects.get('StringNumber')! as StringNumberContainerGlyph;
            }
            container.addString(n.string);
        }
        if (n.isPercussion) {
            const articulation = PercussionMapper.getArticulation(n);
            if (articulation && articulation.techniqueSymbolPlacement !== TextBaseline.Middle) {
                const effectContainer =
                    articulation.techniqueSymbolPlacement === TextBaseline.Top
                        ? this.noteHeads!.aboveBeatEffects
                        : this.noteHeads!.belowBeatEffects;

                switch (articulation.techniqueSymbol) {
                    case MusicFontSymbol.PictEdgeOfCymbal:
                        effectContainer.set('PictEdgeOfCymbal', new PictEdgeOfCymbalGlyph(0, 0));
                        break;
                    case MusicFontSymbol.ArticStaccatoAbove:
                        effectContainer.set('ArticStaccatoAbove', new ArticStaccatoAboveGlyph(0, 0));
                        break;
                    case MusicFontSymbol.StringsUpBow:
                        effectContainer.set('StringsUpBow', new PickStrokeGlyph(0, 0, PickStroke.Up));
                        break;
                    case MusicFontSymbol.StringsDownBow:
                        effectContainer.set('StringsDownBow', new PickStrokeGlyph(0, 0, PickStroke.Down));
                        break;
                    case MusicFontSymbol.GuitarGolpe:
                        effectContainer.set('GuitarGolpe', new GuitarGolpeGlyph(0, 0));
                        break;
                }
            }
        }
    }
}
